ownCloudAppShared/AppLock/AppLockManager.swift

```diff
@@ -417,54 +435,122 @@ public class AppLockManager: NSObject {
+		var shouldDisplayCountdown = false
+
+		if let lockedUntilDate {
+			if determineIfTimeHasBeenTampered(andResetIf: true) {
+				shouldDisplayCountdown = true
+			} else {
+				shouldDisplayCountdown = lockedUntilDate > Date()
+			}
+
+			if !shouldDisplayCountdown {
+				removeLockCountdown()
+			}
+		}
+
+		return shouldDisplayCountdown
+	}
+
+	// MARK: - Countdown
+	private func resetAndStartLockCountdown() {
+		let currentDate = Date()
+		let currentSystemUptime = ProcessInfo.processInfo.systemUptime
+
+		lockedSinceDate = currentDate
+		lockedSinceSystemUptime = currentSystemUptime
+
+		lockedUntilDate = currentDate.addingTimeInterval(lockTimeoutDuration)
+
+		showOrHideLockCountdownAsNeeded()
+	}
+
+	private func removeLockCountdown() {
+		lockedSinceDate = nil
+		lockedSinceSystemUptime = nil
+		lockedUntilDate = nil
+
+		showOrHideLockCountdownAsNeeded()
+	}
+
+	private var timeHasBeenTamperedWith: Bool {
+		// Check for noticable discrepancies between the progress of the clock time and system up time
+		// to mitigate a circumvention tactic that would quit the app, set the date to a distant future date
+		// and launch the app again, allowing the user to have another (sooner) opportunity to try a passcode.
+		let currentDate = Date()
+		let currentSystemUptime = ProcessInfo.processInfo.systemUptime
+
+		guard let lockedSinceDate, let lockedSinceSystemUptime else { return false }
+
+		let clockTimePassedSinceLockStart = currentDate.timeIntervalSince(lockedSinceDate) // Time that passed in clock time since the lock started
+		let systemUptimePassedSinceLockStart = currentSystemUptime - lockedSinceSystemUptime // Time that passed in system uptime since the lock started
+		let differenceBetweenPassedClockTimeAndSystemUptime = abs(clockTimePassedSinceLockStart - systemUptimePassedSinceLockStart) // Discrepancy between time passed in clock time and system uptime
+		let allowedCorridor: TimeInterval = 5 // Maximum allowed discrepancy between the two time intervals
+
+		if differenceBetweenPassedClockTimeAndSystemUptime > allowedCorridor {
+			// Discrepancy higher than allowed corridor => system time appears to have changed
+			return true
 		}
 
 		return false
 	}
 
-	// MARK: - Countdown display
-	private func startLockCountdown() {
-		if self.shouldDisplayCountdown {
+	private func determineIfTimeHasBeenTampered(andResetIf reset: Bool = true) -> Bool {
+		if timeHasBeenTamperedWith {
+			if reset {
+				resetAndStartLockCountdown()
+			}
+			return true
+		}
+		return false
+	}
+
+	// MARK: - UI updates
+	private func showOrHideLockCountdownAsNeeded() {
+		if shouldDisplayCountdown {
 			performPasscodeViewControllerUpdates { (passcodeViewController) in
 				passcodeViewController.keypadButtonsHidden = true
 				passcodeViewController.view.setNeedsLayout()
 			}
+
 			updateLockCountdown()
 
 			lockTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(self.updateLockCountdown), userInfo: nil, repeats: true)
```

Comment:
`lockTimer` needs to be checked if `nil` to prevent duplicate init of the timer. This was also a bug in the previous implementation. could lead in multiple timers.

```suggestion
if lockTimer == nil {
				lockTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(self.updateLockCountdown), userInfo: nil, repeats: true)
			}
```

Reply:
Good catch! I recall timers stopping to fire under certain outside circumstances, so I think re-creating the timer here can't hurt. But the previous timer should definitely be stopped and dropped from memory. To achieve this, I added a `lockTimer?.invalidate()` before `lockTimer = Timer.scheduledTimer(…)`.